sensor/common/admissioncontroller/alert_handler.go

```diff
@@ -46,8 +61,12 @@ func (h *alertHandlerImpl) run() {
 	<-h.stopSig.Done()
 }
 
-func (h *alertHandlerImpl) ProcessAlerts(alerts *sensor.AdmissionControlAlerts) {
+func (h *alertHandlerImpl) ProcessAlerts(alerts *sensor.AdmissionControlAlerts) error {
+	if !h.centralReady.IsDone() {
+		return errCentralNoReachable
+	}
 	go h.processAlerts(alerts)
```

Comment:
Just to confirm: is there any problems if `h.processAlerts` is called when in offline mode? Because state could change at any point of the execution.

Reply:
If the connection breaks while we are processing the alerts we will lose these alerts and central won’t know about them. This is happening in the current implementation. I don’t think a context will help that much in this scenario unless we buffer in sensor. The reason for this is that at this point AC has already deleted the alerts from the buffer because we replied with no error. Canceling `ProcessAlerts` will only ensure we don’t attempt to send the alerts to central, but if we don’t save them somewhere we will lose them regardless. I think the only way to ensure we never lose these alerts is implementing ACK messages between sensor <-> AC (and sensor <-> central or implement a smart way to ensure we don’t lose messages in the [central sender](https://github.com/stackrox/stackrox/blob/170c65580b483db1d30f05ba77738ea8ea41c5ff/sensor/common/sensor/central_sender_impl.go#L81C4-L99)). AC will only remove the alerts from the buffer if it receives the ACK from sensor.

I think it's ok to leave this as it is for v1. This is a very extreme case (connection breaks while the alerts are being forwarded to central) and we are not making things worse than the current implementation.